File: fsm.py
#!/usr/bin/env python3

import sys

def generate(name): #generate dictionary
    state = {}
    readme = open(name)

    here = False
    for line in readme:
        if(here):
            lits = line[line.index(":")+1:].split(",")
            initial =  line[:line.index(":")]
            state[initial]={}
            for combo in lits:
                transition = combo.strip()[:combo.strip().index(":")] 
                final =  combo.strip()[combo.strip().index(":")+1:]
                if transition in state[initial]: # see if transform is not determinstic
                    print("error: nondeterministic transition from "+initial+" upon input "+transition) # example6.1 anything 
                    sys.exit()
                else:
                    state[initial][transition]=final
        elif(line.strip()[:5] == "start"):
            start = line.split()[len(line.split())-1]
        elif(line.strip()[:3] == "end"):
            end = line.split()[len(line.split())-1]
        elif(line=="[transitions]\n"): #the initial part is not entirely helpful
            here=True
    readme.close()
    return (start,end,state)

def tryy(triple,sequence):
    transitions = triple[2]
    state=triple[0]
    for letter in sequence:
        if letter in transitions[state].keys():
            state=transitions[state][letter]
        elif "*" in transitions[state].keys():
            state=transitions[state]["*"]
        else:
            print("error: machine not fully defined") #example4 abd
            sys.exit()
    if(state== triple[1]):
        print("Accepting state")
    else:
        print("Reject. Final state is "+state )
def main():
    tryy(generate(sys.argv[1]),sys.argv[2])
    print(generate(sys.argv[1]))
if __name__=="__main__":
    main()

File: fsm.c
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>


typedef struct dictNode{
    char *key;
    void *value;
    bool accepting;
    struct dictNode *next;
} dictNode;

dictNode *newDictNode(char *me, bool accept){
    dictNode *toreturn = (dictNode *)malloc(sizeof(dictNode));
    toreturn->key=me;
    toreturn->value=NULL;
    toreturn->accepting=accept;
    toreturn->next=NULL;
    return toreturn;
}
void main(int argc, char **argv){
    dictNode *head = (dictNode *) malloc(sizeof(dictNode));
    dictNode **transitions = &head; // to keep track of dictionary for access throughout main
    FILE *readme = fopen(argv[1], "r");
    char here = false;
    char *line=fgets(readme);
    do{

        if(here){
            int lits= 0;
            while(line[lits]!=':'){
                lits++;
            } 
            char *initial = strndup(line,lits);
            dictNode *newNode = newDictNode(initial,false);
            int lits2=lits;
            while (line[lits2]!='0'){
                while(line[lits2] != ':'){
                    lits2++;
                } 
                int lits3=lits2;
                while(line[lits3] != ','){
                    lits3++;
                }
		char *transition = strndup(&initial[lits],lits2-lits);
                char *final = strndup(&initial[lits2],lits3-lits); 
                //transitions[initial][transition]=final;
            }
        }
        elif(line.strip()[:5] == "start"){
            state = line.split()[len(line.split())-1];
        }
        elif(line.strip()[:3] == "end"){
            line.split()[len(line.split())-1];
            //accepting=true;
        }
        elif(strcmp(line,"[transitions]\n")==0){
            here=true;
        }
        line=fgets(readme);
    }while(line!=NULL);
    
    fclose(readme);
    line=argv[2];
    int count=0;
    for(int i=0;i<sizeof(line);i++){
        if letter in transitions[state].keys(){
                    state=transitions[state][letter]
        }        
        elif "*" in transitions[state].keys(){     
               state=transitions[state]["*"]
        }        
    if(state== argv[1]):
        printf("Accepting state");
    else{
        printf("Reject. Final state is %s ", state );
    }
}
